yarn-project/pxe/src/synchronizer/synchronizer.ts

```diff
@@ -13,14 +17,30 @@ import { type PxeDatabase } from '../database/index.js';
  * details, and fetch transactions by hash. The Synchronizer ensures that it maintains the note processors
  * in sync with the blockchain while handling retries and errors gracefully.
  */
-export class Synchronizer {
-  private runningPromise?: RunningPromise;
+export class Synchronizer implements L2BlockStreamEventHandler {
   private running = false;
   private initialSyncBlockNumber = INITIAL_L2_BLOCK_NUM - 1;
   private log: DebugLogger;
+  private blockStream: L2BlockStream;
 
-  constructor(private node: AztecNode, private db: PxeDatabase, private jobQueue: SerialQueue, logSuffix = '') {
+  constructor(private node: AztecNode, private db: PxeDatabase, private l2TipsStore: L2TipsStore, logSuffix = '') {
     this.log = createDebugLogger(logSuffix ? `aztec:pxe_synchronizer_${logSuffix}` : 'aztec:pxe_synchronizer');
+    this.blockStream = new L2BlockStream(node, this.l2TipsStore, this);
+  }
+
+  /** Handle events emitted by the block stream. */
+  async handleBlockStreamEvent(event: L2BlockStreamEvent): Promise<void> {
+    await this.l2TipsStore.handleBlockStreamEvent(event);
+
+    switch (event.type) {
+      case 'blocks-added':
+        await this.db.setHeader(event.blocks.at(-1)!.header);
+        break;
+      case 'chain-pruned':
+        await this.db.unnullifyNotesAfter(event.blockNumber);
+        await this.db.removeNotesAfter(event.blockNumber);
```

Comment:
```suggestion
        // We first unnullify and then remove so that unnullified notes that were created after the block number
        // end up deleted.
        await this.db.unnullifyNotesAfter(event.blockNumber);
        await this.db.removeNotesAfter(event.blockNumber);
```